# Preparing data for modeling

### Importing Important Libraries-

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

## 1. Data preparation for train set-

In [2]:
# loading train dataset
train_dataset = pd.read_csv('Data after Feature Engineering/train_FA.csv')
# setting max_columns to 100, so it displays all the columns of train dataset
pd.set_option('display.max_columns', 100)
# displaying head (top 5) rows of train dataset
train_dataset.head()

msno  is_churn  city    bd  gender  \
0  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
1  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
2  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
3  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
4  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   

   registered_via  registration_init_time  payment_method_id  \
0             3.0     1387756800000000000                0.0   
1             3.0     1387756800000000000                0.0   
2             3.0     1387756800000000000                0.0   
3             3.0     1387756800000000000                0.0   
4             3.0     1387756800000000000                0.0   

   payment_plan_days  plan_list_price  actual_amount_paid  is_auto_renew  \
0               30.0            149.0               149.0            2.0   
1               30.0            149.0               149.0            2.0   
2               30.0            149.0               149.0            2.0   
3               30.0            149.0               149.0            2.0   
4               30.0            149.0               149.0            2.0   

  transaction_date membership_expire_date  is_cancel                 date  \
0       2017-03-16             2017-04-19        2.0  1488672000000000000   
1       2017-03-16             2017-04-19        2.0  1488326400000000000   
2       2017-03-16             2017-04-19        2.0  1489881600000000000   
3       2017-03-16             2017-04-19        2.0  1489622400000000000   
4       2017-03-16             2017-04-19        2.0  1489104000000000000   

   num_25  num_50  num_75  num_985  num_100  num_unq  total_secs  \
0     7.0     0.0     3.0      0.0     71.0     68.0     17600.0   
1     2.0     0.0     0.0      1.0     21.0     16.0      8830.0   
2     0.0     0.0     0.0      0.0     34.0     17.0      7884.0   
3    15.0     0.0     0.0      1.0     38.0     17.0      9030.0   
4     0.0     0.0     0.0      0.0      8.0      8.0      1870.0   

  day_of_the_week  is_weekend  is_weekday  num_25_sum  num_50_sum  num_75_sum  \
0          Sunday           1           0        50.0         4.0         6.0   
1       Wednesday           0           1        50.0         4.0         6.0   
2          Sunday           1           0        50.0         4.0         6.0   
3        Thursday           0           1        50.0         4.0         6.0   
4          Friday           0           1        50.0         4.0         6.0   

   num_985_sum  num_100_sum  num_unq_sum  total_secs_sum  is_weekend_sum  \
0          5.0        246.0        206.0       64499.977               5   
1          5.0        246.0        206.0       64499.977               5   
2          5.0        246.0        206.0       64499.977               5   
3          5.0        246.0        206.0       64499.977               5   
4          5.0        246.0        206.0       64499.977               5   

   is_weekday_sum  num_25_mean  num_50_mean  num_75_mean  num_985_mean  \
0               6        4.547       0.3635       0.5454        0.4546   
1               6        4.547       0.3635       0.5454        0.4546   
2               6        4.547       0.3635       0.5454        0.4546   
3               6        4.547       0.3635       0.5454        0.4546   
4               6        4.547       0.3635       0.5454        0.4546   

   num_100_mean  num_unq_mean  total_secs_mean  num_25_std  num_50_std  \
0         22.36         18.73           5864.0       4.734       0.481   
1         22.36         18.73           5864.0       4.734       0.481   
2         22.36         18.73           5864.0       4.734       0.481   
3         22.36         18.73           5864.0       4.734       0.481   
4         22.36         18.73           5864.0       4.734       0.481   

   num_7

In [3]:
# removing all non_useful features first
train_dataset = train_dataset.drop(['registration_init_time',
                              'transaction_date', 
                              'membership_expire_date', 
                              'date', 
                              'num_25',
                              'num_50',
                              'num_75',
                              'num_985',
                              'num_100',
                              'num_unq',
                              'total_secs',
                              'is_weekend',
                              'is_weekday',
                              'day_of_the_week', 
                              'date_min', 
                              'date_max', 
                              'transaction_date_max', 
                              'membership_expire_date_max'], axis=1, inplace=False)

In [4]:
# getting shape of train dataset
train_dataset.shape

(16887877, 58)

**Now this train dataset has 16887877 rows and 58 columns.**

In [5]:
# adding registration init time feature to train dataset
# load the registration array
reg_date = np.load('Preprocessed Data/train_registration_date.npz')
# display the array
reg_date['arr_0']

array([20131223., 20131223., 20131223., ..., 20110905., 20110905.,
       20150204.])

In [6]:
# converting numpy array to list
registration_init_time = reg_date['arr_0'].tolist()
# getting length of the list
len(registration_init_time)

16887877

In [7]:
# creating new feature, and adding it to train dataset
train_dataset['registration_init_time'] = registration_init_time

In [8]:
# setting max_columns to 100, so it displays all the columns of train dataset
pd.set_option('display.max_columns', 100)
# displaying head (top 5) rows of train dataset
train_dataset.head()

msno  is_churn  city    bd  gender  \
0  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
1  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
2  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
3  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
4  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   

   registered_via  payment_method_id  payment_plan_days  plan_list_price  \
0             3.0                0.0               30.0            149.0   
1             3.0                0.0               30.0            149.0   
2             3.0                0.0               30.0            149.0   
3             3.0                0.0               30.0            149.0   
4             3.0                0.0               30.0            149.0   

   actual_amount_paid  is_auto_renew  is_cancel  num_25_sum  num_50_sum  \
0               149.0            2.0        2.0        50.0         4.0   
1               149.0            2.0        2.0        50.0         4.0   
2               149.0            2.0        2.0        50.0         4.0   
3               149.0            2.0        2.0        50.0         4.0   
4               149.0            2.0        2.0        50.0         4.0   

   num_75_sum  num_985_sum  num_100_sum  num_unq_sum  total_secs_sum  \
0         6.0          5.0        246.0        206.0       64499.977   
1         6.0          5.0        246.0        206.0       64499.977   
2         6.0          5.0        246.0        206.0       64499.977   
3         6.0          5.0        246.0        206.0       64499.977   
4         6.0          5.0        246.0        206.0       64499.977   

   is_weekend_sum  is_weekday_sum  num_25_mean  num_50_mean  num_75_mean  \
0               5               6        4.547       0.3635       0.5454   
1               5               6        4.547       0.3635       0.5454   
2               5               6        4.547       0.3635       0.5454   
3               5               6        4.547       0.3635       0.5454   
4               5               6        4.547       0.3635       0.5454   

   num_985_mean  num_100_mean  num_unq_mean  total_secs_mean  num_25_std  \
0        0.4546         22.36         18.73           5864.0       4.734   
1        0.4546         22.36         18.73           5864.0       4.734   
2        0.4546         22.36         18.73           5864.0       4.734   
3        0.4546         22.36         18.73           5864.0       4.734   
4        0.4546         22.36         18.73           5864.0       4.734   

   num_50_std  num_75_std  num_985_std  num_100_std  num_unq_std  \
0       0.481      0.8906       0.8906        19.42        16.14   
1       0.481      0.8906       0.8906        19.42        16.14   
2       0.481      0.8906       0.8906        19.42        16.14   
3       0.481      0.8906       0.8906        19.42        16.14   
4       0.481      0.8906       0.8906        19.42        16.14   

   total_secs_std  active_days  payment_plan_days_mean  plan_list_price_mean  \
0          4748.0           11                    30.0                 149.0   
1          4748.0           11                    30.0                 149.0   
2          4748.0           11                    30.0                 149.0   
3          4748.0           11                    30.0                 149.0   
4          4748.0           11                    30.0                 149.0   

   actual_amount_paid_mean  transaction_date_count  membership_expire_count  \
0                    149.0                       1                        1   
1                    149.0                       1                        1   
2                    149.0                       1                        1   
3                    149.0                       1                        1   
4               

In [9]:
# saving new dataset
train_dataset.to_csv('Prepare Data/prepared_train.csv', index=False)

In [2]:
# reading csv file for removing duplicates
train_dataset = pd.read_csv('Prepare Data/prepared_train.csv')
# getting all columns of train dataset
train_dataset.columns

Index(['msno', 'is_churn', 'city', 'bd', 'gender', 'registered_via',
       'payment_method_id', 'payment_plan_days', 'plan_list_price',
       'actual_amount_paid', 'is_auto_renew', 'is_cancel', 'num_25_sum',
       'num_50_sum', 'num_75_sum', 'num_985_sum', 'num_100_sum', 'num_unq_sum',
       'total_secs_sum', 'is_weekend_sum', 'is_weekday_sum', 'num_25_mean',
       'num_50_mean', 'num_75_mean', 'num_985_mean', 'num_100_mean',
       'num_unq_mean', 'total_secs_mean', 'num_25_std', 'num_50_std',
       'num_75_std', 'num_985_std', 'num_100_std', 'num_unq_std',
       'total_secs_std', 'active_days', 'payment_plan_days_mean',
       'plan_list_price_mean', 'actual_amount_paid_mean',
       'transaction_date_count', 'membership_expire_count', 'activity_period',
       'inactive_days', 'is_rare', 'avg_time_perday', 'unq_track_perday',
       'till_25_perday', 'till_50_perday', 'till_75_perday', 'till_985_perday',
       'till_full_perday', 'discount', 'is_discount', 'days_since_last_l

In [3]:
# removing all duplicate rows for a particular id
train_dataset = train_dataset.drop_duplicates('msno', keep='first', inplace=False)
# setting max_columns to 100, so it displays all the columns of train dataset
pd.set_option('display.max_columns', 100)
# displaying head (top 10) rows of train dataset
train_dataset.head(10)

msno  is_churn  city    bd  \
0    ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0   
11   f/NmvEzHfhINFEYZTR05prUdr+E+3+oewvweYz9cCQE=         1  13.0  20.0   
17   zLo9f73nGGT1p21ltZC3ChiRnAVvgibMyazbCxvWPcg=         1  13.0  18.0   
57   8iF/+8HY8lJKFrTc7iR9ZYGCG2Ecrogbc2Vy5YhsfhQ=         1   1.0  28.0   
67   K6fja4+jmoZ5xG6BypqX80Uw/XKpMgrEMdG2edFOxnA=         1  13.0  35.0   
187  ibIHVYBqxGwrSExE63/omeDD99M5vYB3CN2HzkEY+eM=         1  22.0  28.0   
191  kVmM8X4iBPCOfK/m1l3KR7LjFoxD1EcKYCc76F5IAWw=         1   1.0  28.0   
192  moRTKhKIDvb+C8ZHOgmaF4dXMLk0jOn65d7a8tQ2Eds=         1   9.0  28.0   
223  dW/tPZMDh2Oz/ksduEctJbsz0MXw3kay/1AlZCq3EbI=         1  13.0  21.0   
281  otEcMhAX3mU4gumUSogqgteN3oaNmhdmTkoof2iRYEE=         1  13.0  28.0   

     gender  registered_via  payment_method_id  payment_plan_days  \
0       1.0             3.0                0.0               30.0   
11      1.0             3.0               36.0               30.0   
17      1.0             3.0               17.0               30.0   
57      0.0             7.0               41.0               30.0   
67      2.0             7.0               41.0               30.0   
187     0.0             3.0               40.0               30.0   
191     0.0             7.0               41.0               30.0   
192     2.0             3.0               38.0               30.0   
223     2.0             9.0               38.0               30.0   
281     0.0             7.0               40.0               30.0   

     plan_list_price  actual_amount_paid  is_auto_renew  is_cancel  \
0              149.0               149.0            2.0        2.0   
11             180.0               180.0            0.0        0.0   
17               0.0                 0.0            0.0        0.0   
57             149.0               149.0            1.0        0.0   
67              99.0                99.0            1.0        0.0   
187            149.0               149.0            1.0        1.0   
191             99.0                99.0            1.0        1.0   
192            149.0               149.0            0.0        0.0   
223            149.0               149.0            0.0        0.0   
281            149.0               149.0            1.0        0.0   

     num_25_sum  num_50_sum  num_75_sum  num_985_sum  num_100_sum  \
0          50.0         4.0         6.0          5.0        246.0   
11          0.0         4.0         2.0          0.0         26.0   
17        150.0        50.0        38.0         30.0        410.0   
57         20.0         0.0         0.0          0.0        140.0   
67         72.0        56.0        32.0         32.0       3928.0   
187         6.0         1.0         1.0          4.0         37.0   
191         2.0         0.0         0.0          0.0         14.0   
192        78.0        20.0        25.0         16.0        812.0   
223       244.0        46.0        36.0         22.0       1186.0   
281       148.0        28.0        22.0         32.0        730.0   

     num_unq_sum  total_secs_sum  is_weekend_sum  is_weekday_sum  num_25_mean  \
0          206.0       64499.977               5               6        4.547   
11          30.0        6986.509               3               3        0.000   
17         732.0      135620.940              16              24        3.750   
57         160.0       38807.650               0              10        2.000   
67        4384.0      992830.000              24              96        0.600   
187         48.0       10110.665               2               2        1.500   
191         16.0        3880.765               0               1        2.000   
192        649.0      221420.770               8              23        2.516   
223       1118.0      319777.620              14              44        4.207   
281        650.0      195195.080               8              42        2.960   

     num_50_mean  num_75_mean  num_9

In [4]:
# getting shape of train dataset
train_dataset.shape

(970960, 59)

**Now this dataset don't have any duplicate rows. And it has 970960 rows and 59 columns**

In [5]:
# getting columns and converting them to list
cols = train_dataset.columns.tolist()
# saving them for further use
np.savez_compressed('Prepare Data/columns.npz', cols)

In [6]:
# saving actual train dataset
train_dataset.to_csv('Prepare Data/actual_train.csv', index=False)

## 2. Data preparation for test data-

In [2]:
# loading test dataset
test_dataset = pd.read_csv('Data after Feature Engineering/test_FA.csv')
# setting max_columns to 100, so it displays all the columns of test dataset
pd.set_option('display.max_columns', 100)
# displaying head (top 5) rows of test dataset
test_dataset.head()

msno  is_churn  city    bd  gender  \
0  4n+fXlyJvfQnTeKXTWT507Ll4JVYGrOC8LHCfwBmPE4=         0   1.0  28.0     0.0   
1  aNmbC1GvFUxQyQUidCVmfbQ0YeCuwkPzEdQ0RwWyeZM=         0   4.0  28.0     1.0   
2  aNmbC1GvFUxQyQUidCVmfbQ0YeCuwkPzEdQ0RwWyeZM=         0   4.0  28.0     1.0   
3  aNmbC1GvFUxQyQUidCVmfbQ0YeCuwkPzEdQ0RwWyeZM=         0   4.0  28.0     1.0   
4  aNmbC1GvFUxQyQUidCVmfbQ0YeCuwkPzEdQ0RwWyeZM=         0   4.0  28.0     1.0   

   registered_via  registration_init_time  payment_method_id  \
0             7.0     1437177600000000000               41.0   
1             9.0     1130630400000000000               34.0   
2             9.0     1130630400000000000               34.0   
3             9.0     1130630400000000000               34.0   
4             9.0     1130630400000000000               34.0   

   payment_plan_days  plan_list_price  actual_amount_paid  is_auto_renew  \
0               30.0             99.0                99.0            1.0   
1               30.0            149.0               149.0            1.0   
2               30.0            149.0               149.0            1.0   
3               30.0            149.0               149.0            1.0   
4               30.0            149.0               149.0            1.0   

  transaction_date membership_expire_date  is_cancel                 date  \
0       2017-03-18             2017-04-18        0.0  1489622400000000000   
1       2017-03-31             2017-04-30        0.0  1488931200000000000   
2       2017-03-31             2017-04-30        0.0  1489881600000000000   
3       2017-03-31             2017-04-30        0.0  1488326400000000000   
4       2017-03-31             2017-04-30        0.0  1488844800000000000   

   num_25  num_50  num_75  num_985  num_100  num_unq  total_secs  \
0     2.0     0.0     0.0      0.0     14.0     16.0      3880.0   
1     2.0     4.0     0.0      0.0     14.0     16.0      3880.0   
2     2.0     2.0     3.0      0.0     34.0     57.0     10340.0   
3    15.0     2.0     0.0      0.0     14.0     16.0      3880.0   
4     3.0     2.0     0.0      1.0     13.0     16.0      3826.0   

  day_of_the_week  is_weekend  is_weekday  num_25_sum  num_50_sum  num_75_sum  \
0        Thursday           0           1         2.0         0.0         0.0   
1       Wednesday           0           1       139.0        44.0        32.0   
2          Sunday           1           0       139.0        44.0        32.0   
3       Wednesday           0           1       139.0        44.0        32.0   
4         Tuesday           0           1       139.0        44.0        32.0   

   num_985_sum  num_100_sum  num_unq_sum  total_secs_sum  is_weekend_sum  \
0          0.0         14.0         16.0        3880.765               0   
1         23.0        803.0        855.0      199222.030               8   
2         23.0        803.0        855.0      199222.030               8   
3         23.0        803.0        855.0      199222.030               8   
4         23.0        803.0        855.0      199222.030               8   

   is_weekday_sum  num_25_mean  num_50_mean  num_75_mean  num_985_mean  \
0               1        2.000        0.000        0.000        0.0000   
1              23        4.484        1.419        1.032        0.7417   
2              23        4.484        1.419        1.032        0.7417   
3              23        4.484        1.419        1.032        0.7417   
4              23        4.484        1.419        1.032        0.7417   

   num_100_mean  num_unq_mean  total_secs_mean  num_25_std  num_50_std  \
0          14.0         16.00           3880.0         NaN         NaN   
1          25.9         27.58           6428.0       4.152       1.501   
2          25.9         27.58           6428.0       4.152       1.501   
3          25.9         27.58           6428.0       4.152       1.501   
4          25.9         27.58           6428.0       4.152       1.501   

   num_7

In [3]:
# removing all non_useful features first
test_dataset = test_dataset.drop(['registration_init_time',
                              'transaction_date', 
                              'membership_expire_date', 
                              'date', 
                              'num_25',
                              'num_50',
                              'num_75',
                              'num_985',
                              'num_100',
                              'num_unq',
                              'total_secs',
                              'is_weekend',
                              'is_weekday',
                              'day_of_the_week', 
                              'date_min', 
                              'date_max', 
                              'transaction_date_max', 
                              'membership_expire_date_max'], axis=1, inplace=False)

In [4]:
# getting shape of test dataset
test_dataset.shape

(13745932, 58)

**Now this test dataset has 13745932 rows and 58 columns.**

In [5]:
# adding registration init time feature to test dataset
# load the registration array
reg_date = np.load('Preprocessed Data/test_registration_date.npz')
# display the array
reg_date['arr_0']

array([20150718., 20051030., 20051030., ..., 20130820., 20130820.,
       20130820.])

In [6]:
# converting numpy array to list
registration_init_time = reg_date['arr_0'].tolist()
# getting length of the list
len(registration_init_time)

13745932

In [7]:
# creating new feature, and adding it to train dataset
test_dataset['registration_init_time'] = registration_init_time

In [8]:
# setting max_columns to 100, so it displays all the columns of test dataset
pd.set_option('display.max_columns', 100)
# displaying head (top 5) rows of test dataset
test_dataset.head()

msno  is_churn  city    bd  gender  \
0  4n+fXlyJvfQnTeKXTWT507Ll4JVYGrOC8LHCfwBmPE4=         0   1.0  28.0     0.0   
1  aNmbC1GvFUxQyQUidCVmfbQ0YeCuwkPzEdQ0RwWyeZM=         0   4.0  28.0     1.0   
2  aNmbC1GvFUxQyQUidCVmfbQ0YeCuwkPzEdQ0RwWyeZM=         0   4.0  28.0     1.0   
3  aNmbC1GvFUxQyQUidCVmfbQ0YeCuwkPzEdQ0RwWyeZM=         0   4.0  28.0     1.0   
4  aNmbC1GvFUxQyQUidCVmfbQ0YeCuwkPzEdQ0RwWyeZM=         0   4.0  28.0     1.0   

   registered_via  payment_method_id  payment_plan_days  plan_list_price  \
0             7.0               41.0               30.0             99.0   
1             9.0               34.0               30.0            149.0   
2             9.0               34.0               30.0            149.0   
3             9.0               34.0               30.0            149.0   
4             9.0               34.0               30.0            149.0   

   actual_amount_paid  is_auto_renew  is_cancel  num_25_sum  num_50_sum  \
0                99.0            1.0        0.0         2.0         0.0   
1               149.0            1.0        0.0       139.0        44.0   
2               149.0            1.0        0.0       139.0        44.0   
3               149.0            1.0        0.0       139.0        44.0   
4               149.0            1.0        0.0       139.0        44.0   

   num_75_sum  num_985_sum  num_100_sum  num_unq_sum  total_secs_sum  \
0         0.0          0.0         14.0         16.0        3880.765   
1        32.0         23.0        803.0        855.0      199222.030   
2        32.0         23.0        803.0        855.0      199222.030   
3        32.0         23.0        803.0        855.0      199222.030   
4        32.0         23.0        803.0        855.0      199222.030   

   is_weekend_sum  is_weekday_sum  num_25_mean  num_50_mean  num_75_mean  \
0               0               1        2.000        0.000        0.000   
1               8              23        4.484        1.419        1.032   
2               8              23        4.484        1.419        1.032   
3               8              23        4.484        1.419        1.032   
4               8              23        4.484        1.419        1.032   

   num_985_mean  num_100_mean  num_unq_mean  total_secs_mean  num_25_std  \
0        0.0000          14.0         16.00           3880.0         NaN   
1        0.7417          25.9         27.58           6428.0       4.152   
2        0.7417          25.9         27.58           6428.0       4.152   
3        0.7417          25.9         27.58           6428.0       4.152   
4        0.7417          25.9         27.58           6428.0       4.152   

   num_50_std  num_75_std  num_985_std  num_100_std  num_unq_std  \
0         NaN         NaN          NaN          NaN          NaN   
1       1.501       1.277        1.095        18.83         17.4   
2       1.501       1.277        1.095        18.83         17.4   
3       1.501       1.277        1.095        18.83         17.4   
4       1.501       1.277        1.095        18.83         17.4   

   total_secs_std  active_days  payment_plan_days_mean  plan_list_price_mean  \
0             NaN            1                    30.0                  99.0   
1          4220.0           31                    30.0                 149.0   
2          4220.0           31                    30.0                 149.0   
3          4220.0           31                    30.0                 149.0   
4          4220.0           31                    30.0                 149.0   

   actual_amount_paid_mean  transaction_date_count  membership_expire_count  \
0                     99.0                       1                        1   
1                    149.0                       1                        1   
2                    149.0                       1                        1   
3                    149.0                       1                        1   
4               

In [9]:
test_dataset.to_csv('Prepare Data/prepared_test.csv', index=False)

In [2]:
# reading csv file for removing duplicates
test_dataset = pd.read_csv('Prepare Data/prepared_test.csv')
# getting all columns of test dataset
test_dataset.columns

Index(['msno', 'is_churn', 'city', 'bd', 'gender', 'registered_via',
       'payment_method_id', 'payment_plan_days', 'plan_list_price',
       'actual_amount_paid', 'is_auto_renew', 'is_cancel', 'num_25_sum',
       'num_50_sum', 'num_75_sum', 'num_985_sum', 'num_100_sum', 'num_unq_sum',
       'total_secs_sum', 'is_weekend_sum', 'is_weekday_sum', 'num_25_mean',
       'num_50_mean', 'num_75_mean', 'num_985_mean', 'num_100_mean',
       'num_unq_mean', 'total_secs_mean', 'num_25_std', 'num_50_std',
       'num_75_std', 'num_985_std', 'num_100_std', 'num_unq_std',
       'total_secs_std', 'active_days', 'payment_plan_days_mean',
       'plan_list_price_mean', 'actual_amount_paid_mean',
       'transaction_date_count', 'membership_expire_count', 'activity_period',
       'inactive_days', 'is_rare', 'avg_time_perday', 'unq_track_perday',
       'till_25_perday', 'till_50_perday', 'till_75_perday', 'till_985_perday',
       'till_full_perday', 'discount', 'is_discount', 'days_since_last_l

In [3]:
# removing all duplicate rows for a particular id
test_dataset = test_dataset.drop_duplicates('msno', keep='first', inplace=False)
# setting max_columns to 100, so it displays all the columns of test dataset
pd.set_option('display.max_columns', 100)
# displaying head (top 10) rows of test dataset
test_dataset.head(10)

msno  is_churn  city    bd  \
0    4n+fXlyJvfQnTeKXTWT507Ll4JVYGrOC8LHCfwBmPE4=         0   1.0  28.0   
1    aNmbC1GvFUxQyQUidCVmfbQ0YeCuwkPzEdQ0RwWyeZM=         0   4.0  28.0   
32   rFC9eSG/tMuzpre6cwcMLZHEYM89xY02qcz7HL4//jc=         0   4.0  34.0   
42   WZ59dLyrQcE7ft06MZ5dj40BnlYQY7PHgg/54+HaCSE=         0   0.0  28.0   
43   aky/Iv8hMp1/V/yQHLtaVuEmmAxkB5GuasQZePJ7NU4=         0   1.0  28.0   
44   nu1jZ/flvSaXRg0U9Es+xmo2KlAXq/q+mhcWFStwm9w=         0  13.0  28.0   
74   biDTtgK83fEWXJDaxzD1eh22dion/h3odeIbS7qJMmY=         0  13.0  26.0   
85   k/8uwi/iM9LZmRAIWXLqpZY6ENomXAscwsQsh6PxcTw=         0   0.0  28.0   
86   18rQ/746SjA6nBx325UsyhfsDhu4tK01FXFxHWZjw20=         0  13.0  32.0   
111  2V13OCoWx6vqKr/ZzNmKFrmnC2FtR4SWMz5C5Hi02PY=         0  14.0  32.0   

     gender  registered_via  payment_method_id  payment_plan_days  \
0       0.0             7.0               41.0               30.0   
1       1.0             9.0               34.0               30.0   
32      1.0             7.0               41.0               30.0   
42      0.0             0.0               41.0               30.0   
43      0.0            13.0               30.0               30.0   
44      0.0             4.0               36.0               30.0   
74      1.0             3.0               37.0               30.0   
85      0.0             0.0               41.0               30.0   
86      1.0             9.0               40.0               30.0   
111     1.0             4.0               40.0               30.0   

     plan_list_price  actual_amount_paid  is_auto_renew  is_cancel  \
0               99.0                99.0            1.0        0.0   
1              149.0               149.0            1.0        0.0   
32              99.0                99.0            1.0        0.0   
42              99.0                99.0            1.0        0.0   
43             129.0               129.0            1.0        0.0   
44             180.0               180.0            0.0        0.0   
74             149.0               149.0            1.0        0.0   
85              99.0                99.0            1.0        0.0   
86             149.0               149.0            1.0        0.0   
111            149.0               149.0            1.0        0.0   

     num_25_sum  num_50_sum  num_75_sum  num_985_sum  num_100_sum  \
0           2.0         0.0         0.0          0.0         14.0   
1         139.0        44.0        32.0         23.0        803.0   
32         14.0         2.0         1.0          2.0         84.0   
42          2.0         0.0         0.0          0.0         14.0   
43          2.0         0.0         0.0          0.0         14.0   
44        108.0        36.0        26.0         25.0        965.0   
74         10.0        10.0         6.0          4.0        173.0   
85          2.0         0.0         0.0          0.0         14.0   
86         17.0         6.0         4.0          6.0        980.0   
111         0.0         0.0         0.0          0.0         18.0   

     num_unq_sum  total_secs_sum  is_weekend_sum  is_weekday_sum  num_25_mean  \
0           16.0        3880.765               0               1        2.000   
1          855.0      199222.030               8              23        4.484   
32          95.0       23814.270               0              10        1.400   
42          16.0        3880.765               0               1        2.000   
43          16.0        3880.765               0               1        2.000   
44        1116.0      251226.750               8              22        3.600   
74         156.0       40830.125               0              11        0.909   
85          16.0        3880.765               0               1        2.000   
86         971.0      248882.840               3              22        0.680   
111         18.0        3997.613               0               3        0.000   

     num_50_mean  num_75_mean  num_9

In [4]:
# getting shape of test dataset
test_dataset.shape

(907471, 59)

**Now this dataset don't have any duplicate rows. And it has 907471 rows and 59 columns**

In [5]:
# saving actual test dataset
test_dataset.to_csv('Prepare Data/actual_test.csv', index=False)

## 3. Train test split-

In [2]:
# loading actual train data
train_data = pd.read_csv('Prepare Data/actual_train.csv')
# setting max_columns to 100, so it displays all the columns of train data
pd.set_option('display.max_columns', 100)
# displaying head (top 5) rows of train data
train_data.head()

msno  is_churn  city    bd  gender  \
0  ugx0CjOMzazClkFzU2xasmDZaoIqOUAZPsH1q0teWCg=         1   5.0  28.0     1.0   
1  f/NmvEzHfhINFEYZTR05prUdr+E+3+oewvweYz9cCQE=         1  13.0  20.0     1.0   
2  zLo9f73nGGT1p21ltZC3ChiRnAVvgibMyazbCxvWPcg=         1  13.0  18.0     1.0   
3  8iF/+8HY8lJKFrTc7iR9ZYGCG2Ecrogbc2Vy5YhsfhQ=         1   1.0  28.0     0.0   
4  K6fja4+jmoZ5xG6BypqX80Uw/XKpMgrEMdG2edFOxnA=         1  13.0  35.0     2.0   

   registered_via  payment_method_id  payment_plan_days  plan_list_price  \
0             3.0                0.0               30.0            149.0   
1             3.0               36.0               30.0            180.0   
2             3.0               17.0               30.0              0.0   
3             7.0               41.0               30.0            149.0   
4             7.0               41.0               30.0             99.0   

   actual_amount_paid  is_auto_renew  is_cancel  num_25_sum  num_50_sum  \
0               149.0            2.0        2.0        50.0         4.0   
1               180.0            0.0        0.0         0.0         4.0   
2                 0.0            0.0        0.0       150.0        50.0   
3               149.0            1.0        0.0        20.0         0.0   
4                99.0            1.0        0.0        72.0        56.0   

   num_75_sum  num_985_sum  num_100_sum  num_unq_sum  total_secs_sum  \
0         6.0          5.0        246.0        206.0       64499.977   
1         2.0          0.0         26.0         30.0        6986.509   
2        38.0         30.0        410.0        732.0      135620.940   
3         0.0          0.0        140.0        160.0       38807.650   
4        32.0         32.0       3928.0       4384.0      992830.000   

   is_weekend_sum  is_weekday_sum  num_25_mean  num_50_mean  num_75_mean  \
0               5               6        4.547       0.3635       0.5454   
1               3               3        0.000       0.6665       0.3333   
2              16              24        3.750       1.2500       0.9500   
3               0              10        2.000       0.0000       0.0000   
4              24              96        0.600       0.4666       0.2666   

   num_985_mean  num_100_mean  num_unq_mean  total_secs_mean  num_25_std  \
0        0.4546        22.360         18.73           5864.0       4.734   
1        0.0000         4.332          5.00           1164.0       0.000   
2        0.7500        10.250         18.30           3390.0       2.980   
3        0.0000        14.000         16.00           3880.0       0.000   
4        0.2666        32.720         36.53           8270.0       0.712   

   num_50_std  num_75_std  num_985_std  num_100_std  num_unq_std  \
0      0.4810      0.8906       0.8906       19.420       16.140   
1      1.1050      0.4714       0.0000        2.926        2.887   
2      1.4795      1.0710       0.7666       12.900       19.250   
3      0.0000      0.0000       0.0000        0.000        0.000   
4      0.8057      0.4421       0.4421       19.060       14.560   

   total_secs_std  active_days  payment_plan_days_mean  plan_list_price_mean  \
0          4748.0           11                    30.0                 149.0   
1           741.5            6                    30.0                 180.0   
2          3882.0           20                    30.0                  74.5   
3             0.0            1                    30.0                 149.0   
4          4612.0           15                    30.0                  99.0   

   actual_amount_paid_mean  transaction_date_count  membership_expire_count  \
0                    149.0                       1                        1   
1                    180.0                       1                        1   
2                     74.5                       2                        2   
3                    149.0                       5                       10   
4               

In [3]:
# checking for null values
train_data.isnull().values.any()

True

In [4]:
# replacing all nan and infinity to 0
train_data = train_data.replace(np.nan, 0, inplace=False)
train_data = train_data.replace(np.inf, 0, inplace=False)

In [5]:
# https://stackoverflow.com/a/29651514
def normalize(df):
    result1 = df.copy()
    for feature_name in df.columns:
        if (str(feature_name) != str('msno') and str(feature_name)!=str('is_churn')):
            max_value = df[feature_name].max()
            min_value = df[feature_name].min()
            result1[feature_name] = (df[feature_name] - min_value) / (max_value - min_value)
    return result1


train_dataset = normalize(train_data)

In [6]:
# getting labels for y
labels = train_dataset['is_churn'].values
labels

array([1, 1, 1, ..., 0, 0, 0])

In [7]:
# getting labels for X
data = train_dataset.drop(['msno', 'is_churn'], axis=1, inplace=False)
data

city        bd  gender  registered_via  payment_method_id  \
0       0.227273  0.385714     0.5        0.230769           0.000000   
1       0.590909  0.271429     0.5        0.230769           0.878049   
2       0.590909  0.242857     0.5        0.230769           0.414634   
3       0.045455  0.385714     0.0        0.538462           1.000000   
4       0.590909  0.485714     1.0        0.538462           1.000000   
...          ...       ...     ...             ...                ...   
970955  0.590909  0.385714     0.5        0.538462           0.975610   
970956  0.045455  0.385714     0.0        0.538462           1.000000   
970957  0.045455  0.385714     0.0        0.538462           1.000000   
970958  0.227273  0.342857     1.0        0.692308           0.951220   
970959  0.000000  0.385714     0.0        0.000000           0.902439   

        payment_plan_days  plan_list_price  actual_amount_paid  is_auto_renew  \
0                     1.0         0.827778            0.827778            1.0   
1                     1.0         1.000000            1.000000            0.0   
2                     1.0         0.000000            0.000000            0.0   
3                     1.0         0.827778            0.827778            0.5   
4                     1.0         0.550000            0.550000            0.5   
...                   ...              ...                 ...            ...   
970955                1.0         0.827778            0.827778            0.5   
970956                1.0         0.550000            0.550000            0.5   
970957                1.0         0.550000            0.550000            0.5   
970958                1.0         0.827778            0.827778            0.5   
970959                1.0         0.827778            0.827778            0.5   

        is_cancel  num_25_sum  num_50_sum  num_75_sum  num_985_sum  \
0             1.0    0.002667    0.001057    0.001744     0.001378   
1             0.0    0.000000    0.001057    0.000581     0.000000   
2             0.0    0.008000    0.013214    0.011047     0.008269   
3             0.0    0.001067    0.000000    0.000000     0.000000   
4             0.0    0.003840    0.014799    0.009302     0.008820   
...           ...         ...         ...         ...          ...   
970955        0.0    0.001707    0.001850    0.001163     0.001378   
970956        0.0    0.001120    0.002114    0.002326     0.001929   
970957        0.0    0.000107    0.000000    0.000000     0.000000   
970958        0.0    0.003520    0.005285    0.003488     0.003308   
970959        0.0    0.000107    0.000000    0.000000     0.000000   

        num_100_sum  num_unq_sum  total_secs_sum  is_weekend_sum  \
0          0.002655     0.001872        0.002772        0.004808   
1          0.000281     0.000265        0.000300        0.002885   
2          0.004425     0.006675        0.005828        0.015385   
3          0.001511     0.001452        0.001668        0.000000   
4          0.042390     0.040023        0.042662        0.023077   
...             ...          ...             ...             ...   
970955     0.002417     0.002155        0.002436        0.006731   
970956     0.000237     0.000484        0.000374        0.000962   
970957     0.000151     0.000137        0.000167        0.000000   
970958     0.007252     0.005634        0.006832        0.005769   
970959     0.000151     0.000137        0.000167        0.000000   

        is_weekday_sum  num_25_mean  num_50_mean  num_75_mean  num_985_mean  \
0             0.001938     0.303133     0.090875     0.181800      0.151533   
1             0.000969     0.000000     0.166625     0.111100      0.000000   
2             0.007752     0.250000     0.312500     0.316667      0.250000   
3             0.003230     0.133333     0.000000     0.000000      0.000000   
4             0.031008     0.040000     0.116650     0.088867      0.088867   
...                ...        

In [8]:
# train test split
X_train, X_cv, y_train, y_cv = train_test_split(data, labels, test_size=0.2, random_state=42)

print(X_train.shape, y_train.shape)
print(X_cv.shape, y_cv.shape)

(776768, 57) (776768,)
(194192, 57) (194192,)


In [9]:
# converting X_train to numpy array
train_arr = X_train.values
train_arr

array([[0.        , 0.38571429, 0.        , ..., 0.55      , 0.03052503,
        0.8445787 ],
       [0.04545455, 0.38571429, 0.        , ..., 0.71666667, 0.        ,
        0.53721041],
       [0.04545455, 0.38571429, 0.        , ..., 0.55      , 0.        ,
        0.69327738],
       ...,
       [0.59090909, 0.31428571, 1.        , ..., 1.        , 0.01953602,
        0.53947025],
       [0.04545455, 0.38571429, 0.        , ..., 0.55      , 0.02319902,
        0.54110747],
       [0.        , 0.38571429, 0.        , ..., 0.71666667, 0.01953602,
        0.8445787 ]])

In [10]:
# checking shape of train array
train_arr.shape

(776768, 57)

In [11]:
# saving it for further use
np.savez_compressed("Prepare Data/actual_train.npz", train_arr)

In [12]:
# displaying y_train
train_labels_arr = y_train
train_labels_arr

array([0, 0, 0, ..., 0, 0, 0])

In [13]:
# checking shape of train labels
train_labels_arr.shape

(776768,)

In [14]:
# saving it for further use
np.savez_compressed("Prepare Data/actual_train_labels.npz", train_labels_arr)

In [15]:
# converting X_cv to numpy array
cv_arr = X_cv.values
cv_arr

array([[0.18181818, 0.42857143, 0.5       , ..., 0.82777778, 0.        ,
        0.61321465],
       [0.        , 0.38571429, 0.        , ..., 0.55      , 0.03052503,
        0.8445787 ],
       [0.68181818, 0.38571429, 0.        , ..., 1.        , 0.02808303,
        0.69100217],
       ...,
       [0.36363636, 0.27142857, 0.5       , ..., 1.        , 0.01098901,
        0.61875663],
       [0.04545455, 0.38571429, 0.        , ..., 0.71666667, 0.01587302,
        0.53646482],
       [0.        , 0.38571429, 0.        , ..., 0.55      , 0.03174603,
        0.8445787 ]])

In [16]:
# checking shape of cv array
cv_arr.shape

(194192, 57)

In [17]:
# saving it for further use
np.savez_compressed("Prepare Data/actual_cv.npz", cv_arr)

In [18]:
# displaying y_cv
cv_labels_arr = y_cv
cv_labels_arr

array([0, 0, 0, ..., 0, 0, 0])

In [19]:
# checking shape of cv labels array
cv_labels_arr.shape

(194192,)

In [20]:
# saving it for further use
np.savez_compressed("Prepare Data/actual_cv_labels.npz", cv_labels_arr)

## 4. Converting test set to numpy array-

In [2]:
# loading actual test data
test_data = pd.read_csv('Prepare Data/actual_test.csv')
# displaying head (top 5) rows of test data
test_data.head()

msno  is_churn  city    bd  gender  \
0  4n+fXlyJvfQnTeKXTWT507Ll4JVYGrOC8LHCfwBmPE4=         0   1.0  28.0     0.0   
1  aNmbC1GvFUxQyQUidCVmfbQ0YeCuwkPzEdQ0RwWyeZM=         0   4.0  28.0     1.0   
2  rFC9eSG/tMuzpre6cwcMLZHEYM89xY02qcz7HL4//jc=         0   4.0  34.0     1.0   
3  WZ59dLyrQcE7ft06MZ5dj40BnlYQY7PHgg/54+HaCSE=         0   0.0  28.0     0.0   
4  aky/Iv8hMp1/V/yQHLtaVuEmmAxkB5GuasQZePJ7NU4=         0   1.0  28.0     0.0   

   registered_via  payment_method_id  payment_plan_days  plan_list_price  \
0             7.0               41.0               30.0             99.0   
1             9.0               34.0               30.0            149.0   
2             7.0               41.0               30.0             99.0   
3             0.0               41.0               30.0             99.0   
4            13.0               30.0               30.0            129.0   

   actual_amount_paid  ...  till_985_perday  till_full_perday  discount  \
0                99.0  ...              0.0              14.0       0.0   
1               149.0  ...              1.0              26.0       0.0   
2                99.0  ...              0.0               8.0       0.0   
3                99.0  ...              0.0              14.0       0.0   
4               129.0  ...              0.0              14.0       0.0   

   is_discount  days_since_last_login  days_left  layality_range  \
0            0                     15         18             609   
1            0                      0         30            4170   
2            0                      3         15             865   
3            0                     15         27             782   
4            0                     15         21              90   

   Perday_price  days_since_final_transaction  registration_init_time  
0      3.300000                            13              20150718.0  
1      4.966667                             0              20051030.0  
2      3.300000                            16              20141101.0  
3      3.300000                             4              20150204.0  
4      4.300000                             9              20161222.0  

[5 rows x 59 columns]

In [3]:
# finding all unique labels
test_data['is_churn'].unique()

array([0])

**Since it is containing only 0, so it's better to not store them.**

In [4]:
# replacing all nan and infinity to 0 (if any)
test_data = test_data.replace(np.nan, 0, inplace=False)
test_data = test_data.replace(np.inf, 0, inplace=False)

In [5]:
# https://stackoverflow.com/a/29651514
def normalize(df):
    result1 = df.copy()
    for feature_name in df.columns:
        if (str(feature_name) != str('msno') and str(feature_name)!=str('is_churn')):
            max_value = df[feature_name].max()
            min_value = df[feature_name].min()
            result1[feature_name] = (df[feature_name] - min_value) / (max_value - min_value)
    return result1


test_dataset = normalize(test_data)

In [6]:
# drop
test_dataset = test_dataset.drop(['msno', 'is_churn'], axis=1, inplace=False)
# df
test_dataset

city        bd  gender  registered_via  payment_method_id  \
0       0.045455  0.385714     0.0        0.538462           1.000000   
1       0.181818  0.385714     0.5        0.692308           0.829268   
2       0.181818  0.471429     0.5        0.538462           1.000000   
3       0.000000  0.385714     0.0        0.000000           1.000000   
4       0.045455  0.385714     0.0        1.000000           0.731707   
...          ...       ...     ...             ...                ...   
907466  0.045455  0.385714     0.0        0.538462           1.000000   
907467  0.045455  0.385714     0.0        0.538462           1.000000   
907468  0.045455  0.385714     0.0        0.538462           1.000000   
907469  0.045455  0.385714     0.0        0.538462           1.000000   
907470  0.045455  0.385714     0.0        0.538462           1.000000   

        payment_plan_days  plan_list_price  actual_amount_paid  is_auto_renew  \
0                     1.0         0.550000            0.550000            0.5   
1                     1.0         0.827778            0.827778            0.5   
2                     1.0         0.550000            0.550000            0.5   
3                     1.0         0.550000            0.550000            0.5   
4                     1.0         0.716667            0.716667            0.5   
...                   ...              ...                 ...            ...   
907466                1.0         0.550000            0.550000            0.5   
907467                1.0         0.550000            0.550000            0.5   
907468                1.0         0.550000            0.550000            0.5   
907469                1.0         0.716667            0.716667            0.5   
907470                1.0         0.550000            0.550000            0.5   

        is_cancel  ...  till_985_perday  till_full_perday  discount  \
0             0.0  ...         0.000000          0.009315       0.0   
1             0.0  ...         0.014493          0.017299       0.0   
2             0.0  ...         0.000000          0.005323       0.0   
3             0.0  ...         0.000000          0.009315       0.0   
4             0.0  ...         0.000000          0.009315       0.0   
...           ...  ...              ...               ...       ...   
907466        0.0  ...         0.000000          0.006653       0.0   
907467        0.0  ...         0.000000          0.009315       0.0   
907468        0.0  ...         0.000000          0.009980       0.0   
907469        0.0  ...         0.014493          0.019295       0.0   
907470        0.0  ...         0.014493          0.009980       0.0   

        is_discount  days_since_last_login  days_left  layality_range  \
0               0.0               0.500000   0.007299        0.134864   
1               0.0               0.000000   0.012452        0.878288   
2               0.0               0.100000   0.006011        0.188309   
3               0.0               0.500000   0.011164        0.170981   
4               0.0               0.500000   0.008587        0.026514   
...             ...                    ...        ...             ...   
907466          0.0               0.133333   0.002576        0.283299   
907467          0.0               0.500000   0.008587        0.448852   
907468          0.0               0.200000   0.007299        0.058455   
907469          0.0               0.000000   0.006870        0.477453   
907470          0.0               0.000000   0.007729        0.280585   

        Perday_price  days_since_final_transaction  registration_init_time  
0           0.550000                      0.015873                0.848530  
1           0.827778                      0.000000                0.082276  
2           0.550000                      0.019536                0.774608  
3           0.550000                      0.004884                0.844579  
4           0.716667                      0.010

In [7]:
# converting df to numpy array
test_dataset = test_dataset.values
test_dataset

array([[0.04545455, 0.38571429, 0.        , ..., 0.55      , 0.01587302,
        0.84852957],
       [0.18181818, 0.38571429, 0.5       , ..., 0.82777778, 0.        ,
        0.08227644],
       [0.18181818, 0.47142857, 0.5       , ..., 0.55      , 0.01953602,
        0.77460837],
       ...,
       [0.04545455, 0.38571429, 0.        , ..., 0.55      , 0.01587302,
        0.9253947 ],
       [0.04545455, 0.38571429, 0.        , ..., 0.71666667, 0.01831502,
        0.53644945],
       [0.04545455, 0.38571429, 0.        , ..., 0.55      , 0.01465201,
        0.69558333]])

In [8]:
# getting shape of numpy array
test_dataset.shape

(907471, 57)

In [ ]:
# saving
np.savez_compressed('Prepare Data/actual_test.npz', test_dataset)